### ISSUE TITLE
Failed to map segment from shared object:

### ISSUE THREAD

**Issue by lidonis** [trakt/Plex-Trakt-Scrobbler#278]
I have this error when i start plex

```
2016-02-20 01:51:25,096 - com.plexapp.plugins.trakttv      (7fa3b2510740) :  CRITICAL (com.plexapp.plugins.trakttv:613) - Exception starting plug-in (most recent call last):
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/core.py", line 606, in start
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 256, in execute
  File "/mnt/data/data/datasys/apps/vht-plex/Library/Application Support/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Code/__init__.py", line 44, in <module>
    import interface
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 333, in __import__
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 44, in load_module
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/loader.py", line 30, in __init__
  File "/mnt/data/data/datasys/apps/vht-plex/Library/Application Support/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Code/interface/__init__.py", line 4, in <module>
    import m_sync
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 333, in __import__
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 44, in load_module
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/loader.py", line 30, in __init__
  File "/mnt/data/data/datasys/apps/vht-plex/Library/Application Support/Plex Media Server/Plug-ins/Trakttv.bundle/Contents/Code/interface/m_sync.py", line 8, in <module>
    from plugin.managers.account import AccountManager
  File "bundles-release/Framework.bundle-dist/Contents/Resources/Versions/2/Python/Framework/code/sandbox.py", line 345, in __import__
ImportError: /mnt/data/data/datasys/apps/vht-plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.plugins.trakttv/Libraries/Linux/x86_64/ucs4/apsw.so: failed to map segment from shared object: Operation not permitted

2016-02-20 01:51:25,097 - plugin.core.logger.handlers.error_reporter.ErrorReporter (7fa3b2510740) :  DEBUG (plugin.core.logger.handlers.error_reporter.ErrorReporter:579) - Sending message of length 1384 to requests+http://sentry.skipthe.net/api/1/store/
```

**Comment by Noiden** [comment]
Same here, but i get this error before that critical one.

I think it's the same tough.

`2016-02-21 21:31:36,890 - plugin.core.libraries.main       (7fdb9a9dd700) :  ERROR (plugin.core.libraries.main:55) - apsw: unavailable - Exception raised in 'test_import': /usr/lib/plexmediaserver/Resources/Plug-in Support/Data/com.plexapp.plugins.trakttv/Libraries/Linux/x86_64/ucs4/apsw.so: failed to map segment from shared object: Operation not permitted
Traceback (most recent call last):
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-7efd046/Trakttv.bundle/Contents/Libraries/Shared/plugin/core/libraries/tests/core/base.py", line 33, in run
    result = func()
  File "/usr/lib/plexmediaserver/Resources/Plug-ins-7efd046/Trakttv.bundle/Contents/Libraries/Shared/plugin/core/libraries/tests/apsw_.py", line 9, in test_import
    import apsw
ImportError: /usr/lib/plexmediaserver/Resources/Plug-in Support/Data/com.plexapp.plugins.trakttv/Libraries/Linux/x86_64/ucs4/apsw.so: failed to map segment from shared`

**Comment by Noiden** [comment]
I should mention that I'm running Plex Media Server on Open Media Vault. So I have symlinked the "Plug-in Support" Directory (because that's in another place in OMV than in /usr/lib/.... to save space on the smaller system disk).

The link is like this,
/usr/lib/plexmediaserver/Resources/Plug-in Support -> /media/a13efef6-cf13-4fd1-912c-9e626298cb16/plexmediaserver/Library/Application Support/Plex Media Server/Plug-in Support/

Don't think it has anything to do with this error, but can be good to know. =)

**Comment by fuzeman** [comment]
I think this error is caused when the native libraries (that are bundled in the plugin) aren't supported by the system.

@lidonis Could you tell me what OS (+ version) you are using and the CPU model or arch (i686, x86_64, arm, etc..) it is running on?

@Noiden Could you tell me what version of Open Media Vault you are using and the CPU model or arch (i686, x86_64, arm, etc..) it is running on?

**Comment by lidonis** [comment]
Ubuntu 15.04 (GNU/Linux 3.19.0-26-generic x86_64)

**Comment by Noiden** [comment]
@fuzeman OMV 2.1.29 Stone Burner x86_64. Kernel Linux 3.2.0-4 amd64

**Comment by fuzeman** [comment]
Thanks, will spin up some virtual machines and do some testing later this week.

**Comment by sorinfatu** [comment]
Can add my system to this too, OMV too:
Release: 2.2
Codename: Stone burner
3.16.0-0.bpo.4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u3~bpo70+1 (2016-01-19) x86_64 GNU/Linux

Plugin path is similar to the other OMV case described

**Comment by Bihod** [comment]
Same error here, glad i've found this.
I have an OMV system too (and the plugin path is similar to the other two cases)

Release:2.2, Stone burner
3.2.0-4-amd64 #1 SMP Debian 3.2.73-2+deb7u2 x86_64

**Comment by sorinfatu** [comment]
not sure what prompted this to not work in a sudden fashion, but, at least a temporary workaround is to disable noexec for the partition that holds the Plug-ins directory

in my case (and probably a lot of OMV cases) the plug-in folder is set on the drive used for storage - issue with it being run from a separate drive and installed from omv extras i guess

the "fix" for this would be to remove "noexec" from fstab for the mount that contains the Plug-ins folder
if running from a usb drive and a single hdd is mapped for storage, the line to be edited in fstab is the one for the hdd

this is not a permanent fix but it will allow the plugin to work again.

hope it helps in troubleshooting the actual issue. the point of failure seems to be apsw.so

**Comment by fuzeman** [comment]
@sorinfatu 
Thanks for the extra information, the issue is that the plugin has some native libraries (apsw, cffi, cryptography, llist) bundled with it now (that require "exec" permission).

The best fix for this might be falling back to using the libraries in your python installation if this error is encountered. This should mean you could install the required libraries with ```pip``` instead and avoid removing the ```noexec``` permission from your mount point .

**Comment by gdolla** [comment]
Ook wonderful thank you for helping
Me. I really appreciate it

**Comment by sorinfatu** [comment]
I tested a few versions, some from a few months back and those gave the same error.
I didn't keep the folder with the working version that i had before the issue manifested but i did not update it before getting the errors (only after).

Considering the libs were present in older snapshots and the issue manifested recentely, it might be related (at least in my case) to one of the updates for OMV.

Not sure if the bundled libs are used just as a fallback if they are not present in the system but this is the last batch of updates i got before the bundle started to error:

`libc-bin:amd64 (2.13-38+deb7u8, 2.13-38+deb7u10), openmediavault:amd64 (2.1.26, 2.2), nginx-full:amd64 (1.2.1-2.2+wheezy3, 1.2.1-2.2+wheezy4), locales:amd64 (2.13-38+deb7u8, 2.13-38+deb7u10), nginx:amd64 (1.2.1-2.2+wheezy3, 1.2.1-2.2+wheezy4), nginx-common:amd64 (1.2.1-2.2+wheezy3, 1.2.1-2.2+wheezy4), cpio:amd64 (2.11+dfsg-0.1+deb7u1, 2.11+dfsg-0.1+deb7u2), multiarch-support:amd64 (2.13-38+deb7u8, 2.13-38+deb7u10), libssh2-1:amd64 (1.4.2-1.1+deb7u1, 1.4.2-1.1+deb7u2), libgcrypt11:amd64 (1.5.0-5+deb7u3, 1.5.0-5+deb7u4), libc6:amd64 (2.13-38+deb7u8, 2.13-38+deb7u10)`

prior to these the bundle worked just fine

**Comment by Noiden** [comment]
@fuzeman Could you write how to install thoose libraries with pip?Just "pip install" on the libraries you listed above? Will your plugin use thoose libraries then, rather den the ones in the data disk? Or do we have to change any setting file in the plugin directory?

**Comment by fuzeman** [comment]
@sorinfatu Thanks for the extra details, will be doing some tests in a VM later today.

@Noiden I still need to make some changes to the plugin to support that, will update this issue with more details when it's ready.

**Comment by Noiden** [comment]
@fuzeman Cool, thanks. I'll stick to the workaround to remove noexec from the fstab for the moment then. =)

**Comment by fuzeman** [comment]
I've implemented some advanced configuration options *(detailed below)* that could be used to resolve this issue, download this version of the plugin: https://github.com/fuzeman/Plex-Trakt-Scrobbler/archive/issue/278.zip

Haven't done much testing yet *(on systems with this issue)*, so let me know if you have any issues getting this working correctly.

-------------------------------

#### Method One

1. Create a new libraries cache directory on your system drive (can be located anywhere on the drive)
    ```
    mkdir -p /var/lib/tfp/libraries
    ```

2. Update ownership (to the user running Plex Media Server)
    ```
    chown plex:plex /var/lib/tfp/libraries
    ```

3. Update permissions
    ```
    chmod -R 770 /var/lib/tfp/libraries
    ```

4. Create the ```Plug-in Support/Data/com.plexapp.plugins.trakttv/advanced.ini``` file with the following content:
    ```
    [libraries]
    cache_path = /var/lib/tfp/libraries
    ```

-------------------------------

#### Method Two

1. Create a new native libraries directory on your system drive (can be located anywhere on the drive)
    ```
    mkdir -p /var/lib/tfp/libraries
    ```

2. Copy native libraries into the directory for your system:
    ```
    cp -r Plug-ins/Trakttv.bundle/Contents/Libraries/Linux /var/lib/tfp/libraries/
    ```

3. Ensure native libraries have the "execute" permission set:
    ```
    find /var/lib/tfp/libraries -name "*.so" -exec chmod +x {} \;
    ```

4. Create the ```Plug-in Support/Data/com.plexapp.plugins.trakttv/advanced.ini``` file with the following content:
    ```
    [libraries]
    libraries_path = /var/lib/tfp/libraries
    ```

**Comment by sorinfatu** [comment]
Method one used and all working fine on OMV so provided the libraries_path flag is followed the second one should work too.

If you need a specific test for that i could do it over the weekend, otherwise the issue is solved by the change.

**Comment by fuzeman** [comment]
@sorinfatu 
Thanks for confirming that worked, I'll include these changes in the next release *(probably next weekend)*

**Comment by fuzeman** [comment]
Changes released in [v1.0.2.0](https://github.com/trakt/Plex-Trakt-Scrobbler/releases/tag/v1.0.2.0)

**Comment by alneven** [comment]
@fuzeman Methode one was also working for me on Openmediavault (OMV). Thanks!
